ENH: introduce a notion of "compatible" stringdtype instances

ngoldbaum opened the pull request:
This is a followup for #26198, it accomplishes the same thing I wanted in that PR.

See the change to the NEP for the details and explanation. 

In short, this relaxes the error checking in stringdtype ufuncs and changes the `common_instance` logic, allowing operations between distinct stringdtype instances as long as the result isn't ambiguous. This makes it much simpler to work with non-default stringdtype instances, since users don't need to zealously convert all ufunc arguments to the same dtype before passing them to numpy in the most common cases like passing a python string as an argument.

For all operations that take more than one string argument, we now only raise an error if the inputs have distinct na_object settings. We allow distinct `coerce` settings and just choose `coerce=False` for string outputs if any input dtype had `coerce=False` set.

Also added a test. There was one spot in the existing tests where we were doing equality comparisons between arrays with distinct na_object settings, so I updated that test to account for the behavior change.

Ideally we could get this merged in time to be included with 2.0 RC2. I'd like to have this in NumPy 2.0 because it will eliminate a lot of boilerplate argument sanitizing in pandas when it called numpy ufuncs. I totally understand if this is coming too late in the game though.

ngoldbaum committed:
MNT: fix copy/paste error for NA type extracted from Pandas

ngoldbaum committed:
ENH: introduce 'compatible' stringdtype instances

mhvk commented on numpy/_core/src/umath/stringdtype_ufuncs.cpp:
```diff
@@ -1303,47 +1222,30 @@ replace_resolve_descriptors(struct PyArrayMethodObject_tag *NPY_UNUSED(method),
     PyArray_Descr *out_descr = NULL;
 
     if (given_descrs[4] == NULL) {
         out_descr = (PyArray_Descr *)new_stringdtype_instance(
```
Why not use `common_instance` directly? It is a new reference already, no? (and DECREF in other branch)

mhvk commented on numpy/_core/src/umath/stringdtype_ufuncs.cpp:
```diff
@@ -1579,42 +1481,28 @@ center_ljust_rjust_resolve_descriptors(
     PyArray_Descr *out_descr = NULL;
 
     if (given_descrs[3] == NULL) {
         out_descr = (PyArray_Descr *)new_stringdtype_instance(
```
As above, why not use `common_instance` directly?

ngoldbaum replied:
That's a good point but it doesn't work because we need to use `common_instance` for the error checking even in the other branch of the enclosing if statement here. Instead I refactored the compatibility checking into its own function and used that, avoiding creating new objects or calling `common_instance`. Thank you for pointing this out, I used `common_instance` without thinking about how it creates a `PyObject` unnecessarily.

mhvk commented on numpy/_core/src/umath/stringdtype_ufuncs.cpp:
```diff
@@ -2646,7 +2542,7 @@ init_stringdtype_ufuncs(PyObject *umath)
     for (int i=0; i<3; i++) {
         if (init_ufunc(umath, strip_chars_names[i], strip_chars_dtypes,
-                       &strip_chars_resolve_descriptors,
+                       &binary_resolve_descriptors,
```
This is really nice!

mhvk commented on numpy/_core/tests/test_stringdtype.py:
```diff
@@ -225,7 +225,22 @@ def test_self_casts(dtype, dtype2, strings):
+        else:
+            # comparisons between arrays with distinct NA objects
+            # aren't allowed
+            if ((na1 is pd_NA or na2 is pd_NA or
```
How about writing 
```
if ((na1 := getattr(dtype, "na_object", None) is not None
    and (na2 := getattr(dtype2, "na_object", None) is not None
    and (na1 is not na2)
    and (na1 is pd_NA or ...)):
    with pytest.raises
        ...
else:
    assert_array_equal(...)
```

ngoldbaum replied:
That doesn't work because `na_object` being unset is the default. It's only a valid attribute if you pass it:

```
>>> np.dtypes.StringDType().na_object
AttributeError: 'numpy.dtypes.StringDType' object has no attribute 'na_object'.
```

but I see what you're saying, refactored a bit to hopefully be less confusing

mhvk reviewed (commented):
commit 1: all OK, but perhaps numpy should consider having some form of pre-commit that checks imports...

commit 2: I guess I understand why `Py_NewRef` doesn't "just work" on array descriptors, but it is a pity that this will force new ref variants for every class one devises. In that sense, the code becomes less readable. But the reduction in boilerplate is probably still worth it.

mhvk reviewed (commented):
Overall, good, but some comments inline - main one being not to adjust `na_cmp_eq` but rather do the check in `common_instance`.

Bigger picture, I wonder a bit about using this inside, e.g., `string_comparison_resolve_descriptors`, since in the end you just don't generally actually use it. Since before no check was done, what goes wrong if you don't adjust this routine? Is it just that one gets a less clear error message? Somewhat similarly with the other resolve descriptors routines where the output of `common_instance` is not used.

seberg commented on numpy/_core/src/multiarray/stringdtype/dtype.c:
```diff
@@ -182,29 +189,29 @@ _eq_comparison(int scoerce, int ocoerce, PyObject *sna, PyObject *ona)
 {
-    int eq = _eq_comparison(dtype1->coerce, dtype2->coerce, dtype1->na_object,
-                            dtype2->na_object);
+    int eq = na_eq_cmp(dtype1->na_object, dtype2->na_object, 1);
```
Seems like a good idea if it isn't used in many places at least.  (I don't care about the perf, the compiler may just inline and optimize it all away anyway.)

seberg commented on numpy/_core/src/multiarray/stringdtype/dtype.h:
```diff
@@ -49,6 +49,9 @@ stringdtype_finalize_descr(PyArray_Descr *dtype);
 _eq_comparison(int scoerce, int ocoerce, PyObject *sna, PyObject *ona);
 
+NPY_NO_EXPORT PyArray_StringDTypeObject *
+common_instance(PyArray_StringDTypeObject *dtype1, PyArray_StringDTypeObject *dtype2);
```
might make sense to name it `string_common_instance` or so if we export it, in case this might end up getting included elsewhere?  (just a passing thought)

seberg reviewed (commented):
Didn't have a very close look, but I think it is good once Marten's comments are addressed.

I like the `NPy_NewDescrRef` cleanup enough that I am happy to add it, although it _would_ be nicer to have a single `NewRef` macro(?) that works for all types...
Maybe even `#define Npy_NewRefT(obj) (Py_NewRef(obj), obj)`, which feels a bit hacky, but not too much.

FWIW, I think we should definitely backport since this is limited to `StringDType` and there is no code (beyond your pandas tries) that even *can* be using it.

ngoldbaum committed:
MNT: refactor stringdtype compatibility checking out of common_instance

ngoldbaum committed:
MNT: refactor tortured logic in test

ngoldbaum commented:
I just pushed an update addressing all your comments.

Regarding `NewRef`, I realized today that using it complicates backports, so I backed out using it in this PR. Sorry for the noise!

Let's not add it until we're no longer actively backporting stuff to numpy 2.0. I guess we could also backport adding pythoncapi-compat, but I'm ok living without `NewRef` for a little while.

ngoldbaum commented on numpy/_core/src/multiarray/stringdtype/dtype.c:
```diff
@@ -185,6 +185,41 @@ _eq_comparison(int scoerce, int ocoerce, PyObject *sna, PyObject *ona)
+// currently this can only return 1 or -1, the latter indicating that the
+// error indicator is set
+NPY_NO_EXPORT int
+stringdtype_compatible_na(PyObject *na1, PyObject *na2) {
```
This is only used by `string_comparison_resolve_descriptors`, in principle I could expose `na_eq_cmp` but exposing this seemed more useful.

mhvk replied:
Agreed this is more useful. My sense, though, would be to make this function like your next one, i.e., have its signature include an `PyObject **out_na`, and, if this is non-NULL, return the compatible na in it. So, something like,
```
NPY_NO_EXPORT int
stringdtype_compatible_na(PyObject *na1, PyObject *na2, PyObject **out_na)
{
    if (na1 != NULL && na2 != NULL) {
        int na_eq = na_eq_comp(na1, na2);
        if (na_eq < 0) {
            return -1;
        }
        else if (na_eq == 0) {
            PyErr_Format(PyExc_TypeError,
                     "Cannot find a compatible null string value for "
                     "null strings '%R' and '%R'", na1, na2);
            return -1;
        }
    }
    if (out_na != NULL) {
        *out_na = na1 ? na1 : na2;
    }
    return 0;
}
```
I think with that, you don't really need the next function any more - in the few places where it is used you can just calculate `coerce` directly.

ngoldbaum replied:
Applied this suggestion in the most recent commit.

ngoldbaum committed:
MNT: refactor stringdtype_setitem following marten's suggestion

ngoldbaum commented on numpy/_core/src/multiarray/stringdtype/dtype.c:
```diff
@@ -280,7 +287,7 @@ stringdtype_setitem(PyArray_StringDTypeObject *descr, PyObject *obj, char **data
     npy_packed_static_string *sdata = (npy_packed_static_string *)dataptr;
 
-    int is_cmp = 0;
+    int na_cmp = 0;
```
Oops, missed this, just pushed another commit addressing this comment. Cleans up the duplicate NULL checking too.

mhvk commented on numpy/_core/src/multiarray/stringdtype/dtype.c:
```diff
@@ -185,6 +185,41 @@ _eq_comparison(int scoerce, int ocoerce, PyObject *sna, PyObject *ona)
+// error indicator is set
+NPY_NO_EXPORT int
+stringdtype_compatible_na(PyObject *na1, PyObject *na2) {
+    if ((na1 == NULL) != (na2 == NULL)) {
```
This should be `((na1 == NULL) || (na2 == NULL))`, right? Both `NULL` is definitely fine.

mhvk reviewed (commented):
The approach makes sense, but a suggestion to just have one added function, which can set an output compatible `na`.

ngoldbaum committed:
MAINT: respond to marten's comments

mhvk commented on numpy/_core/src/multiarray/stringdtype/dtype.c:
```diff
@@ -193,18 +217,18 @@ _eq_comparison(int scoerce, int ocoerce, PyObject *sna, PyObject *ona)
-                "Cannot find common instance for unequal dtype instances");
+    if (stringdtype_compatible_na(
+                dtype1->na_object, dtype2->na_object, &out_na_object) == -1) {
+        PyErr_Format(PyExc_TypeError,
```
One of these I never knows, but isn't the old `ValueError` more appropriate than the new `TypeError`? After all, the type is OK, it is the value of `object_na` that is not.

(Also, I assume an already set error is just overridden? I haven't done much error stuff in C)

ngoldbaum replied:
It's a type error because it's a problem with the dtype for an operation. It would be a ValueError if it was a problem with the array values, not the dtype.

mhvk commented on numpy/_core/src/multiarray/stringdtype/dtype.c:
```diff
@@ -185,6 +185,30 @@ _eq_comparison(int scoerce, int ocoerce, PyObject *sna, PyObject *ona)
+    if (out_na != NULL) {
+        *out_na = na1 ? na1 : na2;
+    }
+    return 1;
```
Shouldn't this be `return 0`? (I.e., success.) I don't think any comparison other than `== -1` is done, so it doesn't affect anything, but 0 seems more logical.

mhvk commented on numpy/_core/tests/test_stringdtype.py:
```diff
@@ -1324,6 +1336,46 @@ def test_strip_ljust_rjust_consistency(string_array, unicode_array):
+    # a dtype instance with an unset na object is compatible
+    # with a dtype that has one set
+
+    # this tests uses the "add" ufunc but all ufuncs that accept more
```
The test actually checks comparisons too, but doesn't really matter - just if for some other reason it is worth pushing another version.

mhvk reviewed (approved):
Thanks, looks good! One nitpick and one question about the exception type, but approving now.

ngoldbaum committed:
MNT: respond to minor comments from marten

ngoldbaum commented:
Thanks so much for your comments!

ngoldbaum commented:
Marten gave this a close look and both Marten and Sebastian were positive about the idea, so I'm going to merge this.
